Symfony Twig component (#201)
glpi-project/docdev: source/devapi/symfonycomponents/index.rst

Symfony Twig Components
=======================

.. versionadded:: 12.0

> Twig Components is a `Symfony UX bundle <https://symfony.com/bundles/ux-twig-component/current/index.html>`_ that allows configuring components in PHP.

Reviewer: ```suggestion
Twig Components is a `Symfony UX bundle <https://symfony.com/bundles/ux-twig-component/current/index.html>`_ allowing to use components in twig template, inspired by html components. It aims to replace twig macros.
```

Reviewer 2: s/twig/Twig/
s/html/HTML/

;)

        <twig:block name="title">
            <h4 class="alert-title">
                Custom title block
            </h4>
            {{ parent() }} {# Renders the parent content — here: "Overridden title" #}
        </twig:block>
    </twig:Alert>

Most components also support a default ``content`` block. To inject content into it, place your markup directly inside the ``<twig:xx>`` tag:

.. code-block:: twig

    <twig:Alert:Danger>
        <twig:block name="title">
            <h2 class="alert-title">
                Custom title block
            </h2>
        </twig:block>

        <div>
            My alert content
        </div>
    </twig:Alert:Danger>

.. image:: /_static/images/symfonycomponents/alert-custom-block.png
   :alt: Example with custom twig block

Component Twig Syntax
^^^^^^^^^^^^^^^^^^^^^

There is also a ``component()`` Twig function, but its use is discouraged except in rare cases. It is less readable and less flexible than the HTML syntax (no block overrides, and it visually blends in with other Twig function calls).

.. code-block:: twig

    {{ component('Alert', {
        type: 'warning',
        title: __('My alert title.')
    }) }}

> This integration mode will not be shown in the component documentation.

Reviewer 2: If its discouraged, maybe just remove this paragraph

Author: It can still be useful for very rare case (but user can probably find about it on symfony doc if they really need it)

So if you think it's still better, I'll remove it

Reviewer 2: No need to rewrite SF docs here; there is no need to document every specific or advanced usage case.
Since this is not specific to GLPI; it can be removed.

Creating a Component
--------------------

`Symfony Documentation <https://symfony.com/bundles/ux-twig-component/current/index.html>`_

A Twig component consists of two parts: a **PHP class** that declares the props and logic, and a **Twig template** that defines the markup.

The PHP Class
^^^^^^^^^^^^^

Create a class under ``src/Twig/Components/`` and annotate it with ``#[AsTwigComponent]``.

By default, **Public properties** become the component's props and are automatically available as variables in the template.

**Public methods** are also accessible from the template via the special ``this`` variable.

.. code-block:: php

>     namespace Twig\Components;

Reviewer 2: For syntx coloration to work:
```suggestion
    <?php
    namespace Twig\Components;
```

The Twig Template
^^^^^^^^^^^^^^^^^

Place templates under ``templates/twig_components/``. Props are available directly as template variables. The component object itself is accessible via ``this``, which is useful for calling methods:

.. code-block:: twig

    <div class="{{ this.computedClass }}">
        {% block title %}
            {% if title|length %}
                <h4>{{ title }}</h4>
            {% endif %}
        {% endblock %}

        {% block content %}{% endblock %}
    </div>

Define ``{% block %}`` sections for any part of the markup that consumers may need to override.

The ``{% block content %}`` block is special: any markup placed directly inside the component tag (without an explicit ``<twig:block>``) is injected into it automatically:

.. code-block:: twig

    <twig:Alert>This text is injected into the content block.</twig:Alert>

Variants
^^^^^^^^

Variant components share a base class and, typically, the same template. The class name determines the component tag name: a class ``Twig\Components\Alert\Danger`` automatically resolves to the tag ``<twig:Alert:Danger>``.

.. code-block:: php

>     namespace Twig\Components\Alert;

Reviewer 2: ```suggestion
    <?php
    namespace Twig\Components\Alert;
```

    use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;

    #[AsTwigComponent(template: 'twig_components/Alert/Info.html.twig')]
    final class Danger extends AbstractAlert
    {
        public string $type = 'danger';
    }

The ``template`` parameter is specified explicitly here because all Alert variants share a single template file (``twig_components/Alert/Info.html.twig``).

Testing
^^^^^^^

Two levels of tests are recommended:

- **Unit tests**: instantiate the PHP class directly and assert prop defaults and method return values. No GLPI environment needed.
- **Rendering tests**: render a Twig string using ``TemplateRenderer::getInstance()->renderFromStringTemplate()`` and assert the resulting HTML. These extend ``GLPITestCase``.

Tests live in ``tests/functional/Twig/Components/`` (`GitHub <https://github.com/glpi-project/glpi/tree/main/tests/functional/Twig/Components/>`_). See ``AlertTest.php`` and ``AlertRenderingTest.php`` for examples.

Debugging
^^^^^^^^^

To list all registered components and their resolved template paths, run:

> .. code-block:: bash

Reviewer 2: ```suggestion
::
```
